// @pulumi/gcp/serviceusage/consumerQuotaOverride.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * A consumer override is applied to the consumer on its own authority to limit its own quota usage.
 * Consumer overrides cannot be used to grant more quota than would be allowed by admin overrides,
 * producer overrides, or the default limit of the service.
 *
 * To get more information about ConsumerQuotaOverride, see:
 *
 * * How-to Guides
 *     * [Getting Started](https://cloud.google.com/service-usage/docs/getting-started)
 *     * [REST API documentation](https://cloud.google.com/service-usage/docs/reference/rest/v1beta1/services.consumerQuotaMetrics.limits.consumerOverrides)
 *
 * ## Example Usage
 * ### Consumer Quota Override
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const myProject = new gcp.organizations.Project("myProject", {
 *     projectId: "quota",
 *     orgId: "123456789",
 * }, {
 *     provider: google_beta,
 * });
 * const override = new gcp.serviceusage.ConsumerQuotaOverride("override", {
 *     project: myProject.projectId,
 *     service: "servicemanagement.googleapis.com",
 *     metric: `servicemanagement.googleapis.com%2Fdefault_requests`,
 *     limit: `%2Fmin%2Fproject`,
 *     overrideValue: "95",
 *     force: true,
 * }, {
 *     provider: google_beta,
 * });
 * ```
 * ### Region Consumer Quota Override
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const myProject = new gcp.organizations.Project("myProject", {
 *     projectId: "quota",
 *     orgId: "123456789",
 * }, {
 *     provider: google_beta,
 * });
 * const override = new gcp.serviceusage.ConsumerQuotaOverride("override", {
 *     dimensions: {
 *         region: "us-central1",
 *     },
 *     project: myProject.projectId,
 *     service: "compute.googleapis.com",
 *     metric: `compute.googleapis.com%2Fn2_cpus`,
 *     limit: `%2Fproject%2Fregion`,
 *     overrideValue: "8",
 *     force: true,
 * }, {
 *     provider: google_beta,
 * });
 * ```
 *
 * ## Import
 *
 * ConsumerQuotaOverride can be imported using any of these accepted formats
 *
 * ```sh
 *  $ pulumi import gcp:serviceusage/consumerQuotaOverride:ConsumerQuotaOverride default projects/{{project}}/services/{{service}}/consumerQuotaMetrics/{{metric}}/limits/{{limit}}/consumerOverrides/{{name}}
 * ```
 *
 * ```sh
 *  $ pulumi import gcp:serviceusage/consumerQuotaOverride:ConsumerQuotaOverride default {{project}}/{{service}}/{{metric}}/{{limit}}/{{name}}
 * ```
 *
 * ```sh
 *  $ pulumi import gcp:serviceusage/consumerQuotaOverride:ConsumerQuotaOverride default {{service}}/{{metric}}/{{limit}}/{{name}}
 * ```
 */
class ConsumerQuotaOverride extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["dimensions"] = state ? state.dimensions : undefined;
            inputs["force"] = state ? state.force : undefined;
            inputs["limit"] = state ? state.limit : undefined;
            inputs["metric"] = state ? state.metric : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["overrideValue"] = state ? state.overrideValue : undefined;
            inputs["project"] = state ? state.project : undefined;
            inputs["service"] = state ? state.service : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.limit === undefined) && !opts.urn) {
                throw new Error("Missing required property 'limit'");
            }
            if ((!args || args.metric === undefined) && !opts.urn) {
                throw new Error("Missing required property 'metric'");
            }
            if ((!args || args.overrideValue === undefined) && !opts.urn) {
                throw new Error("Missing required property 'overrideValue'");
            }
            if ((!args || args.service === undefined) && !opts.urn) {
                throw new Error("Missing required property 'service'");
            }
            inputs["dimensions"] = args ? args.dimensions : undefined;
            inputs["force"] = args ? args.force : undefined;
            inputs["limit"] = args ? args.limit : undefined;
            inputs["metric"] = args ? args.metric : undefined;
            inputs["overrideValue"] = args ? args.overrideValue : undefined;
            inputs["project"] = args ? args.project : undefined;
            inputs["service"] = args ? args.service : undefined;
            inputs["name"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(ConsumerQuotaOverride.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing ConsumerQuotaOverride resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ConsumerQuotaOverride(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ConsumerQuotaOverride.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ConsumerQuotaOverride.__pulumiType;
    }
}
exports.ConsumerQuotaOverride = ConsumerQuotaOverride;
/** @internal */
ConsumerQuotaOverride.__pulumiType = 'gcp:serviceusage/consumerQuotaOverride:ConsumerQuotaOverride';
