// @pulumi/gcp/tags/tagValue.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * A TagValue is a child of a particular TagKey. TagValues are used to group cloud resources for the purpose of controlling them using policies.
 *
 * To get more information about TagValue, see:
 *
 * * [API documentation](https://cloud.google.com/resource-manager/reference/rest/v3/tagValues)
 * * How-to Guides
 *     * [Official Documentation](https://cloud.google.com/resource-manager/docs/tags/tags-creating-and-managing)
 *
 * ## Example Usage
 * ### Tag Value Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const key = new gcp.tags.TagKey("key", {
 *     description: "For keyname resources.",
 *     parent: "organizations/123456789",
 *     shortName: "keyname",
 * });
 * const value = new gcp.tags.TagValue("value", {
 *     description: "For valuename resources.",
 *     parent: pulumi.interpolate`tagKeys/${key.name}`,
 *     shortName: "valuename",
 * });
 * ```
 *
 * ## Import
 *
 * TagValue can be imported using any of these accepted formats
 *
 * ```sh
 *  $ pulumi import gcp:tags/tagValue:TagValue default tagValues/{{name}}
 * ```
 *
 * ```sh
 *  $ pulumi import gcp:tags/tagValue:TagValue default {{name}}
 * ```
 */
class TagValue extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["createTime"] = state ? state.createTime : undefined;
            inputs["description"] = state ? state.description : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["namespacedName"] = state ? state.namespacedName : undefined;
            inputs["parent"] = state ? state.parent : undefined;
            inputs["shortName"] = state ? state.shortName : undefined;
            inputs["updateTime"] = state ? state.updateTime : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.parent === undefined) && !opts.urn) {
                throw new Error("Missing required property 'parent'");
            }
            if ((!args || args.shortName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'shortName'");
            }
            inputs["description"] = args ? args.description : undefined;
            inputs["parent"] = args ? args.parent : undefined;
            inputs["shortName"] = args ? args.shortName : undefined;
            inputs["createTime"] = undefined /*out*/;
            inputs["name"] = undefined /*out*/;
            inputs["namespacedName"] = undefined /*out*/;
            inputs["updateTime"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(TagValue.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing TagValue resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new TagValue(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of TagValue.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === TagValue.__pulumiType;
    }
}
exports.TagValue = TagValue;
/** @internal */
TagValue.__pulumiType = 'gcp:tags/tagValue:TagValue';
