// @pulumi/gcp/containeranalysis/occurence.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * An occurrence is an instance of a Note, or type of analysis that
 * can be done for a resource.
 *
 * To get more information about Occurrence, see:
 *
 * * [API documentation](https://cloud.google.com/container-analysis/api/reference/rest/)
 * * How-to Guides
 *     * [Official Documentation](https://cloud.google.com/container-analysis/)
 *
 * ## Example Usage
 *
 * ## Import
 *
 * Occurrence can be imported using any of these accepted formats
 *
 * ```sh
 *  $ pulumi import gcp:containeranalysis/occurence:Occurence default projects/{{project}}/occurrences/{{name}}
 * ```
 *
 * ```sh
 *  $ pulumi import gcp:containeranalysis/occurence:Occurence default {{project}}/{{name}}
 * ```
 *
 * ```sh
 *  $ pulumi import gcp:containeranalysis/occurence:Occurence default {{name}}
 * ```
 */
class Occurence extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["attestation"] = state ? state.attestation : undefined;
            inputs["createTime"] = state ? state.createTime : undefined;
            inputs["kind"] = state ? state.kind : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["noteName"] = state ? state.noteName : undefined;
            inputs["project"] = state ? state.project : undefined;
            inputs["remediation"] = state ? state.remediation : undefined;
            inputs["resourceUri"] = state ? state.resourceUri : undefined;
            inputs["updateTime"] = state ? state.updateTime : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.attestation === undefined) && !opts.urn) {
                throw new Error("Missing required property 'attestation'");
            }
            if ((!args || args.noteName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'noteName'");
            }
            if ((!args || args.resourceUri === undefined) && !opts.urn) {
                throw new Error("Missing required property 'resourceUri'");
            }
            inputs["attestation"] = args ? args.attestation : undefined;
            inputs["noteName"] = args ? args.noteName : undefined;
            inputs["project"] = args ? args.project : undefined;
            inputs["remediation"] = args ? args.remediation : undefined;
            inputs["resourceUri"] = args ? args.resourceUri : undefined;
            inputs["createTime"] = undefined /*out*/;
            inputs["kind"] = undefined /*out*/;
            inputs["name"] = undefined /*out*/;
            inputs["updateTime"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(Occurence.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Occurence resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Occurence(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Occurence.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Occurence.__pulumiType;
    }
}
exports.Occurence = Occurence;
/** @internal */
Occurence.__pulumiType = 'gcp:containeranalysis/occurence:Occurence';
