// @pulumi/gcp/cloudasset/projectFeed.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Describes a Cloud Asset Inventory feed used to to listen to asset updates.
 *
 * To get more information about ProjectFeed, see:
 *
 * * [API documentation](https://cloud.google.com/asset-inventory/docs/reference/rest/)
 * * How-to Guides
 *     * [Official Documentation](https://cloud.google.com/asset-inventory/docs)
 *
 * ## Example Usage
 *
 * ## Import
 *
 * ProjectFeed can be imported using any of these accepted formats
 *
 * ```sh
 *  $ pulumi import gcp:cloudasset/projectFeed:ProjectFeed default projects/{{project}}/feeds/{{name}}
 * ```
 *
 * ```sh
 *  $ pulumi import gcp:cloudasset/projectFeed:ProjectFeed default {{project}}/{{name}}
 * ```
 *
 * ```sh
 *  $ pulumi import gcp:cloudasset/projectFeed:ProjectFeed default {{name}}
 * ```
 */
class ProjectFeed extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["assetNames"] = state ? state.assetNames : undefined;
            inputs["assetTypes"] = state ? state.assetTypes : undefined;
            inputs["billingProject"] = state ? state.billingProject : undefined;
            inputs["condition"] = state ? state.condition : undefined;
            inputs["contentType"] = state ? state.contentType : undefined;
            inputs["feedId"] = state ? state.feedId : undefined;
            inputs["feedOutputConfig"] = state ? state.feedOutputConfig : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["project"] = state ? state.project : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.feedId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'feedId'");
            }
            if ((!args || args.feedOutputConfig === undefined) && !opts.urn) {
                throw new Error("Missing required property 'feedOutputConfig'");
            }
            inputs["assetNames"] = args ? args.assetNames : undefined;
            inputs["assetTypes"] = args ? args.assetTypes : undefined;
            inputs["billingProject"] = args ? args.billingProject : undefined;
            inputs["condition"] = args ? args.condition : undefined;
            inputs["contentType"] = args ? args.contentType : undefined;
            inputs["feedId"] = args ? args.feedId : undefined;
            inputs["feedOutputConfig"] = args ? args.feedOutputConfig : undefined;
            inputs["project"] = args ? args.project : undefined;
            inputs["name"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(ProjectFeed.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing ProjectFeed resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ProjectFeed(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ProjectFeed.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ProjectFeed.__pulumiType;
    }
}
exports.ProjectFeed = ProjectFeed;
/** @internal */
ProjectFeed.__pulumiType = 'gcp:cloudasset/projectFeed:ProjectFeed';
