// @pulumi/gcp/diagflow/intent.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Represents a Dialogflow intent. Intents convert a number of user expressions or patterns into an action. An action
 * is an extraction of a user command or sentence semantics.
 *
 * To get more information about Intent, see:
 *
 * * [API documentation](https://cloud.google.com/dialogflow/docs/reference/rest/v2/projects.agent.intents)
 * * How-to Guides
 *     * [Official Documentation](https://cloud.google.com/dialogflow/docs/)
 *
 * ## Example Usage
 * ### Dialogflow Intent Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const basicAgent = new gcp.diagflow.Agent("basicAgent", {
 *     displayName: "example_agent",
 *     defaultLanguageCode: "en",
 *     timeZone: "America/New_York",
 * });
 * const basicIntent = new gcp.diagflow.Intent("basicIntent", {displayName: "basic-intent"}, {
 *     dependsOn: [basicAgent],
 * });
 * ```
 *
 * ## Import
 *
 * Intent can be imported using any of these accepted formats
 *
 * ```sh
 *  $ pulumi import gcp:diagflow/intent:Intent default {{name}}
 * ```
 */
class Intent extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["action"] = state ? state.action : undefined;
            inputs["defaultResponsePlatforms"] = state ? state.defaultResponsePlatforms : undefined;
            inputs["displayName"] = state ? state.displayName : undefined;
            inputs["events"] = state ? state.events : undefined;
            inputs["followupIntentInfos"] = state ? state.followupIntentInfos : undefined;
            inputs["inputContextNames"] = state ? state.inputContextNames : undefined;
            inputs["isFallback"] = state ? state.isFallback : undefined;
            inputs["mlDisabled"] = state ? state.mlDisabled : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["parentFollowupIntentName"] = state ? state.parentFollowupIntentName : undefined;
            inputs["priority"] = state ? state.priority : undefined;
            inputs["project"] = state ? state.project : undefined;
            inputs["resetContexts"] = state ? state.resetContexts : undefined;
            inputs["rootFollowupIntentName"] = state ? state.rootFollowupIntentName : undefined;
            inputs["webhookState"] = state ? state.webhookState : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.displayName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'displayName'");
            }
            inputs["action"] = args ? args.action : undefined;
            inputs["defaultResponsePlatforms"] = args ? args.defaultResponsePlatforms : undefined;
            inputs["displayName"] = args ? args.displayName : undefined;
            inputs["events"] = args ? args.events : undefined;
            inputs["inputContextNames"] = args ? args.inputContextNames : undefined;
            inputs["isFallback"] = args ? args.isFallback : undefined;
            inputs["mlDisabled"] = args ? args.mlDisabled : undefined;
            inputs["parentFollowupIntentName"] = args ? args.parentFollowupIntentName : undefined;
            inputs["priority"] = args ? args.priority : undefined;
            inputs["project"] = args ? args.project : undefined;
            inputs["resetContexts"] = args ? args.resetContexts : undefined;
            inputs["webhookState"] = args ? args.webhookState : undefined;
            inputs["followupIntentInfos"] = undefined /*out*/;
            inputs["name"] = undefined /*out*/;
            inputs["rootFollowupIntentName"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(Intent.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Intent resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Intent(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Intent.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Intent.__pulumiType;
    }
}
exports.Intent = Intent;
/** @internal */
Intent.__pulumiType = 'gcp:diagflow/intent:Intent';
