// @pulumi/gcp/dataloss/preventionDeidentifyTemplate.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Allows creation of templates to de-identify content.
 *
 * To get more information about DeidentifyTemplate, see:
 *
 * * [API documentation](https://cloud.google.com/dlp/docs/reference/rest/v2/projects.deidentifyTemplates)
 * * How-to Guides
 *     * [Official Documentation](https://cloud.google.com/dlp/docs/concepts-templates)
 *
 * ## Example Usage
 * ### Dlp Deidentify Template Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const basic = new gcp.dataloss.PreventionDeidentifyTemplate("basic", {
 *     deidentifyConfig: {
 *         infoTypeTransformations: {
 *             transformations: [
 *                 {
 *                     infoTypes: [
 *                         {
 *                             name: "PHONE_NUMBER",
 *                         },
 *                         {
 *                             name: "CREDIT_CARD_NUMBER",
 *                         },
 *                     ],
 *                     primitiveTransformation: {
 *                         replaceConfig: {
 *                             newValue: {
 *                                 integerValue: 9,
 *                             },
 *                         },
 *                     },
 *                 },
 *                 {
 *                     infoTypes: [
 *                         {
 *                             name: "EMAIL_ADDRESS",
 *                         },
 *                         {
 *                             name: "LAST_NAME",
 *                         },
 *                     ],
 *                     primitiveTransformation: {
 *                         characterMaskConfig: {
 *                             charactersToIgnores: [{
 *                                 commonCharactersToIgnore: "PUNCTUATION",
 *                             }],
 *                             maskingCharacter: "X",
 *                             numberToMask: 4,
 *                             reverseOrder: true,
 *                         },
 *                     },
 *                 },
 *                 {
 *                     infoTypes: [{
 *                         name: "DATE_OF_BIRTH",
 *                     }],
 *                     primitiveTransformation: {
 *                         replaceConfig: {
 *                             newValue: {
 *                                 dateValue: {
 *                                     day: 1,
 *                                     month: 1,
 *                                     year: 2020,
 *                                 },
 *                             },
 *                         },
 *                     },
 *                 },
 *             ],
 *         },
 *     },
 *     description: "Description",
 *     displayName: "Displayname",
 *     parent: "projects/my-project-name",
 * });
 * ```
 *
 * ## Import
 *
 * DeidentifyTemplate can be imported using any of these accepted formats
 *
 * ```sh
 *  $ pulumi import gcp:dataloss/preventionDeidentifyTemplate:PreventionDeidentifyTemplate default {{parent}}/deidentifyTemplates/{{name}}
 * ```
 *
 * ```sh
 *  $ pulumi import gcp:dataloss/preventionDeidentifyTemplate:PreventionDeidentifyTemplate default {{parent}}/{{name}}
 * ```
 */
class PreventionDeidentifyTemplate extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["deidentifyConfig"] = state ? state.deidentifyConfig : undefined;
            inputs["description"] = state ? state.description : undefined;
            inputs["displayName"] = state ? state.displayName : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["parent"] = state ? state.parent : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.deidentifyConfig === undefined) && !opts.urn) {
                throw new Error("Missing required property 'deidentifyConfig'");
            }
            if ((!args || args.parent === undefined) && !opts.urn) {
                throw new Error("Missing required property 'parent'");
            }
            inputs["deidentifyConfig"] = args ? args.deidentifyConfig : undefined;
            inputs["description"] = args ? args.description : undefined;
            inputs["displayName"] = args ? args.displayName : undefined;
            inputs["parent"] = args ? args.parent : undefined;
            inputs["name"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(PreventionDeidentifyTemplate.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing PreventionDeidentifyTemplate resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new PreventionDeidentifyTemplate(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of PreventionDeidentifyTemplate.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === PreventionDeidentifyTemplate.__pulumiType;
    }
}
exports.PreventionDeidentifyTemplate = PreventionDeidentifyTemplate;
/** @internal */
PreventionDeidentifyTemplate.__pulumiType = 'gcp:dataloss/preventionDeidentifyTemplate:PreventionDeidentifyTemplate';
