// @pulumi/gcp/certificateauthority/certificate.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * A Certificate corresponds to a signed X.509 certificate issued by a CertificateAuthority.
 *
 * > **Note:** The Certificate Authority that is referenced by this resource **must** be
 * `tier = "ENTERPRISE"`
 *
 * > **Warning:** Please remember that all resources created during preview (via this provider)
 * will be deleted when CA service transitions to General Availability (GA). Relying on these
 * certificate authorities for production traffic is discouraged.
 *
 * ## Example Usage
 * ### Privateca Certificate Csr
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 * import * from "fs";
 *
 * const test_ca = new gcp.certificateauthority.Authority("test-ca", {
 *     certificateAuthorityId: "my-certificate-authority",
 *     location: "us-central1",
 *     tier: "ENTERPRISE",
 *     config: {
 *         subjectConfig: {
 *             subject: {
 *                 organization: "HashiCorp",
 *             },
 *             commonName: "my-certificate-authority",
 *             subjectAltName: {
 *                 dnsNames: ["hashicorp.com"],
 *             },
 *         },
 *         reusableConfig: {
 *             reusableConfig: "projects/568668481468/locations/us-central1/reusableConfigs/root-unconstrained",
 *         },
 *     },
 *     keySpec: {
 *         algorithm: "RSA_PKCS1_4096_SHA256",
 *     },
 *     disableOnDelete: true,
 * }, {
 *     provider: google_beta,
 * });
 * const _default = new gcp.certificateauthority.Certificate("default", {
 *     project: "my-project-name",
 *     location: "us-central1",
 *     certificateAuthority: test_ca.certificateAuthorityId,
 *     lifetime: "860s",
 *     pemCsr: fs.readFileSync("test-fixtures/rsa_csr.pem"),
 * }, {
 *     provider: google_beta,
 * });
 * ```
 *
 * ## Import
 *
 * Certificate can be imported using any of these accepted formats
 *
 * ```sh
 *  $ pulumi import gcp:certificateauthority/certificate:Certificate default projects/{{project}}/locations/{{location}}/certificateAuthorities/{{certificate_authority}}/certificates/{{name}}
 * ```
 *
 * ```sh
 *  $ pulumi import gcp:certificateauthority/certificate:Certificate default {{project}}/{{location}}/{{certificate_authority}}/{{name}}
 * ```
 *
 * ```sh
 *  $ pulumi import gcp:certificateauthority/certificate:Certificate default {{location}}/{{certificate_authority}}/{{name}}
 * ```
 */
class Certificate extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["certificateAuthority"] = state ? state.certificateAuthority : undefined;
            inputs["certificateDescriptions"] = state ? state.certificateDescriptions : undefined;
            inputs["config"] = state ? state.config : undefined;
            inputs["createTime"] = state ? state.createTime : undefined;
            inputs["labels"] = state ? state.labels : undefined;
            inputs["lifetime"] = state ? state.lifetime : undefined;
            inputs["location"] = state ? state.location : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["pemCertificate"] = state ? state.pemCertificate : undefined;
            inputs["pemCertificates"] = state ? state.pemCertificates : undefined;
            inputs["pemCsr"] = state ? state.pemCsr : undefined;
            inputs["project"] = state ? state.project : undefined;
            inputs["revocationDetails"] = state ? state.revocationDetails : undefined;
            inputs["updateTime"] = state ? state.updateTime : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.certificateAuthority === undefined) && !opts.urn) {
                throw new Error("Missing required property 'certificateAuthority'");
            }
            if ((!args || args.location === undefined) && !opts.urn) {
                throw new Error("Missing required property 'location'");
            }
            inputs["certificateAuthority"] = args ? args.certificateAuthority : undefined;
            inputs["config"] = args ? args.config : undefined;
            inputs["labels"] = args ? args.labels : undefined;
            inputs["lifetime"] = args ? args.lifetime : undefined;
            inputs["location"] = args ? args.location : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["pemCsr"] = args ? args.pemCsr : undefined;
            inputs["project"] = args ? args.project : undefined;
            inputs["certificateDescriptions"] = undefined /*out*/;
            inputs["createTime"] = undefined /*out*/;
            inputs["pemCertificate"] = undefined /*out*/;
            inputs["pemCertificates"] = undefined /*out*/;
            inputs["revocationDetails"] = undefined /*out*/;
            inputs["updateTime"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(Certificate.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Certificate resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Certificate(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Certificate.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Certificate.__pulumiType;
    }
}
exports.Certificate = Certificate;
/** @internal */
Certificate.__pulumiType = 'gcp:certificateauthority/certificate:Certificate';
