// @pulumi/gcp/healthcare/fhirStore.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * A FhirStore is a datastore inside a Healthcare dataset that conforms to the FHIR (https://www.hl7.org/fhir/STU3/)
 * standard for Healthcare information exchange
 *
 * To get more information about FhirStore, see:
 *
 * * [API documentation](https://cloud.google.com/healthcare/docs/reference/rest/v1/projects.locations.datasets.fhirStores)
 * * How-to Guides
 *     * [Creating a FHIR store](https://cloud.google.com/healthcare/docs/how-tos/fhir)
 *
 * ## Example Usage
 * ### Healthcare Fhir Store Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const topic = new gcp.pubsub.Topic("topic", {});
 * const dataset = new gcp.healthcare.Dataset("dataset", {location: "us-central1"});
 * const _default = new gcp.healthcare.FhirStore("default", {
 *     dataset: dataset.id,
 *     version: "R4",
 *     enableUpdateCreate: false,
 *     disableReferentialIntegrity: false,
 *     disableResourceVersioning: false,
 *     enableHistoryImport: false,
 *     notificationConfig: {
 *         pubsubTopic: topic.id,
 *     },
 *     labels: {
 *         label1: "labelvalue1",
 *     },
 * });
 * ```
 * ### Healthcare Fhir Store Streaming Config
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const dataset = new gcp.healthcare.Dataset("dataset", {location: "us-central1"});
 * const bqDataset = new gcp.bigquery.Dataset("bqDataset", {
 *     datasetId: "bq_example_dataset",
 *     friendlyName: "test",
 *     description: "This is a test description",
 *     location: "US",
 *     deleteContentsOnDestroy: true,
 * });
 * const _default = new gcp.healthcare.FhirStore("default", {
 *     dataset: dataset.id,
 *     version: "R4",
 *     enableUpdateCreate: false,
 *     disableReferentialIntegrity: false,
 *     disableResourceVersioning: false,
 *     enableHistoryImport: false,
 *     labels: {
 *         label1: "labelvalue1",
 *     },
 *     streamConfigs: [{
 *         resourceTypes: ["Observation"],
 *         bigqueryDestination: {
 *             datasetUri: pulumi.interpolate`bq://${bqDataset.project}.${bqDataset.datasetId}`,
 *             schemaConfig: {
 *                 recursiveStructureDepth: 3,
 *             },
 *         },
 *     }],
 * });
 * const topic = new gcp.pubsub.Topic("topic", {});
 * ```
 *
 * ## Import
 *
 * FhirStore can be imported using any of these accepted formats
 *
 * ```sh
 *  $ pulumi import gcp:healthcare/fhirStore:FhirStore default {{dataset}}/fhirStores/{{name}}
 * ```
 *
 * ```sh
 *  $ pulumi import gcp:healthcare/fhirStore:FhirStore default {{dataset}}/{{name}}
 * ```
 */
class FhirStore extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["dataset"] = state ? state.dataset : undefined;
            inputs["disableReferentialIntegrity"] = state ? state.disableReferentialIntegrity : undefined;
            inputs["disableResourceVersioning"] = state ? state.disableResourceVersioning : undefined;
            inputs["enableHistoryImport"] = state ? state.enableHistoryImport : undefined;
            inputs["enableUpdateCreate"] = state ? state.enableUpdateCreate : undefined;
            inputs["labels"] = state ? state.labels : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["notificationConfig"] = state ? state.notificationConfig : undefined;
            inputs["selfLink"] = state ? state.selfLink : undefined;
            inputs["streamConfigs"] = state ? state.streamConfigs : undefined;
            inputs["version"] = state ? state.version : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.dataset === undefined) && !opts.urn) {
                throw new Error("Missing required property 'dataset'");
            }
            inputs["dataset"] = args ? args.dataset : undefined;
            inputs["disableReferentialIntegrity"] = args ? args.disableReferentialIntegrity : undefined;
            inputs["disableResourceVersioning"] = args ? args.disableResourceVersioning : undefined;
            inputs["enableHistoryImport"] = args ? args.enableHistoryImport : undefined;
            inputs["enableUpdateCreate"] = args ? args.enableUpdateCreate : undefined;
            inputs["labels"] = args ? args.labels : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["notificationConfig"] = args ? args.notificationConfig : undefined;
            inputs["streamConfigs"] = args ? args.streamConfigs : undefined;
            inputs["version"] = args ? args.version : undefined;
            inputs["selfLink"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(FhirStore.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing FhirStore resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new FhirStore(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of FhirStore.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === FhirStore.__pulumiType;
    }
}
exports.FhirStore = FhirStore;
/** @internal */
FhirStore.__pulumiType = 'gcp:healthcare/fhirStore:FhirStore';
