// @pulumi/gcp/datacatalog/policyTag.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Denotes one policy tag in a taxonomy.
 *
 * To get more information about PolicyTag, see:
 *
 * * [API documentation](https://cloud.google.com/data-catalog/docs/reference/rest/v1beta1/projects.locations.taxonomies.policyTags)
 * * How-to Guides
 *     * [Official Documentation](https://cloud.google.com/data-catalog/docs)
 *
 * ## Example Usage
 * ### Data Catalog Taxonomies Policy Tag Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const myTaxonomy = new gcp.datacatalog.Taxonomy("myTaxonomy", {
 *     region: "us",
 *     displayName: "taxonomy_display_name",
 *     description: "A collection of policy tags",
 *     activatedPolicyTypes: ["FINE_GRAINED_ACCESS_CONTROL"],
 * }, {
 *     provider: google_beta,
 * });
 * const basicPolicyTag = new gcp.datacatalog.PolicyTag("basicPolicyTag", {
 *     taxonomy: myTaxonomy.id,
 *     displayName: "Low security",
 *     description: "A policy tag normally associated with low security items",
 * }, {
 *     provider: google_beta,
 * });
 * ```
 * ### Data Catalog Taxonomies Policy Tag Child Policies
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as gcp from "@pulumi/gcp";
 *
 * const myTaxonomy = new gcp.datacatalog.Taxonomy("myTaxonomy", {
 *     region: "us",
 *     displayName: "taxonomy_display_name",
 *     description: "A collection of policy tags",
 *     activatedPolicyTypes: ["FINE_GRAINED_ACCESS_CONTROL"],
 * }, {
 *     provider: google_beta,
 * });
 * const parentPolicy = new gcp.datacatalog.PolicyTag("parentPolicy", {
 *     taxonomy: myTaxonomy.id,
 *     displayName: "High",
 *     description: "A policy tag category used for high security access",
 * }, {
 *     provider: google_beta,
 * });
 * const childPolicy = new gcp.datacatalog.PolicyTag("childPolicy", {
 *     taxonomy: myTaxonomy.id,
 *     displayName: "ssn",
 *     description: "A hash of the users ssn",
 *     parentPolicyTag: parentPolicy.id,
 * }, {
 *     provider: google_beta,
 * });
 * const childPolicy2 = new gcp.datacatalog.PolicyTag("childPolicy2", {
 *     taxonomy: myTaxonomy.id,
 *     displayName: "dob",
 *     description: "The users date of birth",
 *     parentPolicyTag: parentPolicy.id,
 * }, {
 *     provider: google_beta,
 *     dependsOn: [childPolicy],
 * });
 * ```
 *
 * ## Import
 *
 * PolicyTag can be imported using any of these accepted formats
 *
 * ```sh
 *  $ pulumi import gcp:datacatalog/policyTag:PolicyTag default {{name}}
 * ```
 */
class PolicyTag extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["childPolicyTags"] = state ? state.childPolicyTags : undefined;
            inputs["description"] = state ? state.description : undefined;
            inputs["displayName"] = state ? state.displayName : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["parentPolicyTag"] = state ? state.parentPolicyTag : undefined;
            inputs["taxonomy"] = state ? state.taxonomy : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.displayName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'displayName'");
            }
            if ((!args || args.taxonomy === undefined) && !opts.urn) {
                throw new Error("Missing required property 'taxonomy'");
            }
            inputs["description"] = args ? args.description : undefined;
            inputs["displayName"] = args ? args.displayName : undefined;
            inputs["parentPolicyTag"] = args ? args.parentPolicyTag : undefined;
            inputs["taxonomy"] = args ? args.taxonomy : undefined;
            inputs["childPolicyTags"] = undefined /*out*/;
            inputs["name"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(PolicyTag.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing PolicyTag resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new PolicyTag(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of PolicyTag.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === PolicyTag.__pulumiType;
    }
}
exports.PolicyTag = PolicyTag;
/** @internal */
PolicyTag.__pulumiType = 'gcp:datacatalog/policyTag:PolicyTag';
