// @pulumi/gcp/iap/brand.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * OAuth brand data. Only "Organization Internal" brands can be created
 * programmatically via API. To convert it into an external brands
 * please use the GCP Console.
 *
 * > **Note:** Brands can only be created once for a Google Cloud
 * project and the underlying Google API doesn't not support DELETE or PATCH methods.
 * Destroying a provider-managed Brand will remove it from state
 * but *will not delete it from Google Cloud.*
 *
 * To get more information about Brand, see:
 *
 * * [API documentation](https://cloud.google.com/iap/docs/reference/rest/v1/projects.brands)
 * * How-to Guides
 *     * [Setting up IAP Brand](https://cloud.google.com/iap/docs/tutorial-gce#set_up_iap)
 *
 * ## Example Usage
 *
 * ## Import
 *
 * Brand can be imported using any of these accepted formats
 *
 * ```sh
 *  $ pulumi import gcp:iap/brand:Brand default {{name}}
 * ```
 */
class Brand extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["applicationTitle"] = state ? state.applicationTitle : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["orgInternalOnly"] = state ? state.orgInternalOnly : undefined;
            inputs["project"] = state ? state.project : undefined;
            inputs["supportEmail"] = state ? state.supportEmail : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.applicationTitle === undefined) && !opts.urn) {
                throw new Error("Missing required property 'applicationTitle'");
            }
            if ((!args || args.supportEmail === undefined) && !opts.urn) {
                throw new Error("Missing required property 'supportEmail'");
            }
            inputs["applicationTitle"] = args ? args.applicationTitle : undefined;
            inputs["project"] = args ? args.project : undefined;
            inputs["supportEmail"] = args ? args.supportEmail : undefined;
            inputs["name"] = undefined /*out*/;
            inputs["orgInternalOnly"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(Brand.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Brand resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Brand(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Brand.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Brand.__pulumiType;
    }
}
exports.Brand = Brand;
/** @internal */
Brand.__pulumiType = 'gcp:iap/brand:Brand';
